src/parser/jobs/sam/Buffs.js

```diff
@@ -0,0 +1,136 @@
+		super(...args)
+
+		const filter = {
+			by: 'player',
+			abilityId: [STATUSES.JINPU.id, STATUSES.SHIFU.id],
+		}
+		this.addHook('applybuff', filter, this._onShifuApplication, this._onJinpuApplication)
+		this.addHook('refreshbuff', filter, this._onShifuApplication, this._onJinpuApplication)
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onJinpuApplication(event) {
+		this._JinpuUses.unshift(event)
+
+		if (this._JinpuUses.length < 2) {
+			return
+		}
+
+		const current = this._JinpuUses[0].timestamp
+		const previous = this._JinpuUses[1].timestamp
+		const timeSinceLastApplication = current - previous
+
+		if (timeSinceLastApplication < JINPU_DURATION - BUFF_BUFFER) {
+			this._earlyJinpuApplications++
+		}
+	}
+
+	_onShifuApplication(event) {
+		this._ShifuUses.unshift(event)
+
+		if (this._ShifuUses.length < 2) {
+			return
+		}
+
+		const current = this._ShifuUses[0].timestamp
+		const previous = this._ShifuUses[1].timestamp
+		const timeSinceLastApplication = current - previous
+
+		if (timeSinceLastApplication < SHIFU_DURATION - BUFF_BUFFER) {
+			this._earlyShifuApplications++
+		}
+	}
+
+	_onComplete() {
+		this.checklist.add(new Rule({
+			name: 'Keep Shifu and Jinpu up',
+			description: 'Jinpu increases your damage by 10% and Shifu increases your speed by 10%. Both buffs are key part of Samurai\'s damage.',
+			target: 95,
+			requirements: [
+				new Requirement({
+					name: <Fragment><ActionLink {...ACTIONS.JINPU} /> uptime</Fragment>,
+					percent: () => this.getJinpuUptimePercent(),
+				}),
+
+				new Requirement({
+					name: <Fragment><ActionLink {...ACTIONS.SHIFU} /> uptime</Fragment>,
+					percent: () => this.getShifuUptimePercent(),
+				}),
+
+			],
+		}))
+
+		if (this._earlyJinpuApplications) {
+			this.suggestions.add(new Suggestion({
+				icon: ACTIONS.JINPU.icon,
+				content: <Fragment>
+						Avoid refreshing {ACTIONS.JINPU.name} significantly before its expiration -- That might be making you possibly lose resources.
+				</Fragment>,
+				severity: SEVERITY.MEDIUM,
+				why: <Fragment>
+					{this._earlyJinpuApplications} reapplications that were {BUFF_BUFFER / 1000} or more seconds before expiration.
+				</Fragment>,
+			}))
+		}
+
+		if (this._earlyShifuApplications) {
+			this.suggestions.add(new Suggestion({
+				icon: ACTIONS.SHIFU.icon,
+				content: <Fragment>
+                                                Avoid refreshing {ACTIONS.SHIFU.name} significantly before its expiration -- That might be making you possibly lose resources.
+				</Fragment>,
+				severity: SEVERITY.MEDIUM,
+				why: <Fragment>
+					{this._earlyShifuApplications} reapplications that were {BUFF_BUFFER / 1000} or more seconds before expiration.
+				</Fragment>,
+			}))
+		}
+
+	}
+
+	//
+	getJinpuUptimePercent() {
+		const statusUptime = this.combatants.getStatusUptime(STATUSES.JINPU.id, this.parser.player.id)
+		const fightUptime = this.parser.fightDuration - this.invuln.getInvulnerableUptime()
+
```

Comment:
You could honestly reduce this from two to one method, by changing it to

`getUptimePercent(statusId)` and passing the `STATUSES.JUNPU.id` as a parameter instead. Same for the `STATUSES.SHIFU.id`.